main.py

```diff
@@ -1,68 +1,225 @@
+
+	To stop the loop, use function set on passed event: event.set()
+	'''
+
+	def __init__(self, event: Event, interval: float):
+		Thread.__init__(self)
+		self.stopped = event
+		self.interval = interval
+
+	def run(self):
+		while not self.stopped.wait(self.interval):
+			save_db()
+
+print('creating saving timer')
+SAVE_INTERVAL = 5 * 60 # 5 minutes
+stop_saving_flag = Event()
+save_timer = SaverThread(stop_saving_flag, SAVE_INTERVAL)
+save_timer.start()
+
+print("reading config")
+CURDIR=path.dirname(argv[0])
+with open(path.join(CURDIR,"./config.json"),"r") as f:
+	CONFIG=json.load(f)
+
+print("initializing commands")
+
+updater=Updater(CONFIG["TOKEN"])
+
+def command(name:str):#some python magic
+	def add_it(func):
+		updater.dispatcher.add_handler(CommandHandler(name, func))
+	return add_it
+
+def message(filters):
+	def add_it(func):
+		updater.dispatcher.add_handler(MessageHandler(filters,func))
+	return add_it
+
+def filter_chat(id, chat):
+	'''
+	id: id of a chat
+	chat: @<chat> without @
+	'''
+	def decorator(function):
+		def wrapper(update: Update, context: CallbackContext):
+			if update.message.chat_id != id:
+				update.message.chat.send_message(f'This feature only works in chat @{chat}')
+				return
+			return function(update, context)
+		return wrapper
+	return decorator
+
+
+@command("ping")
+def ping(update: Update, context: CallbackContext):
+	dt=datetime.now(update.message.date.tzinfo)-update.message.date
+	update.message.reply_text(f'Ping is {dt.total_seconds():.2f}s')
+
+def escape_md(txt: str) -> str:
+	return escape_markdown(txt, 2)
+
+def get_mention(user: User):
+	return user.mention_markdown_v2()
+
+@message(Filters.status_update.new_chat_members)
+def new_chat_member(update: Update, context: CallbackContext):
+	handles=", ".join(get_mention(user) for user in update.message.new_chat_members)
+	update.message.reply_text(
+f"""{handles},
+いらっしゃいませ\\! \\[Welcome\\!\\]
+Welcome to this chat\\! Please read the rules\\.
+Добро пожаловать в чат\\! Прочти правила, пожалуйста\\.
+このチャットへようこそ！ ルールをお読みください。
+
+[rules](https://t\\.me/dev\\_meme/3667)""",
+parse_mode=ParseMode.MARKDOWN_V2)
+
+def is_admin(chat: Chat, user: User) -> bool:
+	# might wanna cache admins
+	status = chat.get_member(user.id).status
+	return status == 'creator' or status == 'administrator'
+
+def get_warn_target(message: Message) -> User | None:
+	'''
+	Returns the user that is supposed to be warned. It might be a bot.
+	Returns None if no warn target.
+	'''
+	if message.reply_to_message is not None and message.reply_to_message.from_user is not None:
+		return message.reply_to_message.from_user
+	return None
+
+class WarnPossibleResult(NamedTuple):
+	'''
+	can_warn -> can warn happen (user is admin, user is not a bot, warn target isnt bot)
+	reason -> reason why warn cannot happen, might be None
+	'''
+	is_possible: bool
+	reason: str | None
+
+def is_warn_possible(message: Message, command: str) -> WarnPossibleResult:

```

Comment:
This can be simplified to 'return `str` if error, return `None` otherwise' - we can remove WarnPossibleResult and a bit of complexity that way.

Reply:
Okay. This was a repeating part, so I was thinking of way to remove that, but still making it readable.

Reply:
Reworked it a bit. Now it returns just bool, and this function is responsible for sending error messages.